src/main/java/org/opensearch/knn/index/query/rescore/RescoreSearcher.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.rescore;
+
+import lombok.AllArgsConstructor;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.Bits;
+import org.opensearch.knn.index.query.KNNWeight;
+import org.opensearch.knn.index.query.ResultUtil;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
+public final class RescoreSearcher {
+
+    private final ResultUtil resultUtil;
+
+    /**
+     * Executes a k-NN query by executing a first pass search over an index with quantized vectors with
+     * an oversample factor and then rescoring down to k results for the whole shard.
+     *
+     * @param indexSearcher The index searcher for the current search
+     * @param leafReaderContexts The leaf reader contexts for the current search
+     * @param knnWeight The knn weight for the current search
+     * @return The top k results for the shard
+     * @throws IOException If an error occurs during the search
+     */
+    public TopDocs[] search(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int firstPassK,
+        int finalK
+    ) throws IOException {
+        List<Map<Integer, Float>> perLeafResults = firstPassSearch(indexSearcher, leafReaderContexts, knnWeight, firstPassK);
+        return rescoreResults(indexSearcher, leafReaderContexts, knnWeight, perLeafResults, finalK);
+    }
+
+    private List<Map<Integer, Float>> firstPassSearch(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int k
+    ) throws IOException {
+        List<Callable<Map<Integer, Float>>> tasks = new ArrayList<>(leafReaderContexts.size());
+        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
+            tasks.add(() -> firstPassSearchLeaf(leafReaderContext, knnWeight));
+        }
```

Comment:
For first pass, why can't we reuse the existing logic in [NativeEngineKnnVectorQuery.java](https://github.com/opensearch-project/k-NN/pull/1984/files#diff-9cfe412357ba56b3ef216427d491fc653535686a760e8ba19ea1aa00fc0e0338). we just have to manipulate the k or is there some performance benefits which I missed?

One thing you will have to take care is you need to subtract docBase per leaf in rescoreResults 
This causes duplication adding on to the maintainance

```
// current knnweight will manipulate k if rescorercontext is in query
List<Callable<TopDocs>> tasks = new ArrayList<>(leafReaderContexts.size());
        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
            tasks.add(() -> searchLeaf(leafReaderContext, knnWeight));
        }
        TopDocs[] perLeafResults = indexSearcher.getTaskExecutor().invokeAll(tasks).toArray(TopDocs[]::new);
        // TopDocs.merge requires perLeafResults to be sorted in descending order.
        TopDocs topK = TopDocs.merge(knnQuery.getK(), perLeafResults);
       TopDocs rescoredResults = rescore context == null? tokK : rescoresearcher.rescoreResults(...)?

  
```

Reply:
I think I figured it, we need access to doc Ids of per leaf results so we get the flat values